Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebKit

cdumez opened the pull request:
#### 6c78d5f3b973b90e15c86c0704655b36c18a9c9a
<pre>
Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=282524">https://bugs.webkit.org/show_bug.cgi?id=282524</a>

Reviewed by Geoffrey Garen.

* Source/WebKit/GPUProcess/EntryPoint/Cocoa/XPCService/GPUServiceEntryPoint.mm:
(GPU_SERVICE_INITIALIZER):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportReceiveStreamCocoa.mm:
(WebKit::NetworkTransportReceiveStream::receiveLoop):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Platform/IPC/DaemonCoders.cpp:
(WebKit::Daemon::Coder&lt;WebCore::CertificateInfo&gt;::encode):
* Source/WebKit/Platform/IPC/DaemonCoders.h:
* Source/WebKit/Platform/IPC/DaemonDecoder.cpp:
* Source/WebKit/Platform/IPC/Decoder.cpp:
* Source/WebKit/Platform/IPC/Encoder.cpp:
* Source/WebKit/Platform/IPC/Encoder.h:
(IPC::Encoder::encodeObject):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendStream):
(IPC::StreamClientConnection::trySendSyncStream):
(IPC::StreamClientConnection::trySendDestinationIDIfNeeded):
(IPC::StreamClientConnection::sendProcessOutOfStreamMessage):
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.h:
(IPC::StreamConnectionBuffer::header const):
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::StreamServerConnection::sendSyncReply):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::getDescriptorAndAdvance):
(IPC::Connection::sendOutgoingMessage):
(IPC::createMessageDecoder):
(IPC::Connection::receiveSourceEventHandler):
* Source/WebKit/Platform/IPC/cocoa/DaemonConnectionCocoa.mm:
* Source/WebKit/Platform/IPC/cocoa/MachMessage.h:
(IPC::MachMessage::span):
* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::StreamConnectionEncoder&gt;::Impl::Impl):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionEncoderTests.cpp:
(TestWebKitAPI::TEST(StreamConnectionEncoderTest, MinimumMessageSizeCanEncodeExpectedMessageAtAnyOffset)):

Canonical link: <a href="https://commits.webkit.org/286137@main">https://commits.webkit.org/286137@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ccbdddad3ecf5ba111d67106749f51e542f6524

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74951 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27768 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79381 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26190 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77068 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63517 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2166 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58846 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17121 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78018 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49005 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64394 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39237 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21893 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24522 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68088 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67449 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22237 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80867 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/74209 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2269 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1373 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67110 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2418 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64413 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66410 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10342 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8505 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/96480 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11563 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2234 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/5022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21085 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2262 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3183 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2269 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -94,7 +93,7 @@ template<typename T, size_t Extent> struct ArgumentCoder<std::span<T, Extent>> {
             return data;
         else
-            return std::span<T, Extent> { data.data(), Extent };
+            return unsafeMakeSpan<T, Extent>(data.data(), Extent);
```
I think you can do data.subspan<0, Extent>() here. Slightly nicer not to use the unsafe function.

geoffreygaren replied:
Or maybe data.first<Extent>() is better? I'm new to span API, lol.

cdumez replied:
Going with `data.subspan<0, Extent>()` which seems to work.

geoffreygaren commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -114,7 +113,7 @@ struct ArgumentCoder<ArrayReferenceTuple<Types...>> {
             return;
 
-        (..., encoder.encodeSpan(std::span(arrayReference.template data<Indices>(), size)));
+        (..., encoder.encodeSpan(unsafeMakeSpan(arrayReference.template data<Indices>(), size)));
```
ArrayReferenceTuple has a span() accessor now. Slightly nicer not to use the unsafe function here.

geoffreygaren commented on Source/WebKit/Platform/IPC/Encoder.cpp:
```diff
@@ -48,10 +47,10 @@ static inline std::span<uint8_t> allocBuffer(size_t size)
 #if OS(DARWIN)
     auto* buffer = static_cast<uint8_t*>(mmap(0, size, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0));
```
Probably not worth it in this patch, but eventually we're going to want to make an mmapSpan wrapper.

darinadler replied:
We should make it have an interface similar to `MallocSpan`.

geoffreygaren commented on Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
```diff
@@ -60,15 +57,17 @@ class StreamConnectionEncoder final {
         auto bytes = asBytes(span);
-        size_t bufferPointer = static_cast<size_t>(reinterpret_cast<intptr_t>(m_buffer + m_encodedSize));
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+        size_t bufferPointer = static_cast<size_t>(reinterpret_cast<intptr_t>(m_buffer.data() + m_encodedSize));
```
I think you can do m_buffer.subspan(m_encodedSize).data() to eliminate the skip markers (and gain a bound check).

cdumez replied:
Yes, this was my initial attempt. However, `m_buffer.subspan(m_encodedSize)` hits the assertion in subspan() so I'm leaving it alone for now.

geoffreygaren reviewed (approved):
r=me, plus I added some comments that might help tighten the warnings even more

darinadler commented on Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
```diff
@@ -60,15 +57,17 @@ class StreamConnectionEncoder final {
-        uint8_t* buffer = m_buffer + alignedSize;
-        memcpy(buffer, bytes.data(), bytes.size());
+        auto buffer = m_buffer.subspan(alignedSize);
+        memcpySpan(buffer, bytes);
```
This would read better without the local variable, I think.

darinadler commented on Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
```diff
@@ -77,7 +76,7 @@ class StreamConnectionEncoder final {
     {
         static_assert(std::is_trivially_copyable_v<T>);
-        return encodeSpan(std::span(std::addressof(object), 1));
+        return encodeSpan(unsafeMakeSpan(std::addressof(object), 1));
```
Should we put a wrapper somewhere that safely makes a span for a single object we have a reference to? Doesn’t seem at all unsafe!

cdumez replied:
```
template<typename T>
std::span<const T> asSingleElementSpan(const T& object)
{
    return unsafeMakeSpan(std::addressof(object), 1);
}
```

Or do you have a suggestion for a better name?

darinadler replied:
I like the name. Although I would just call it `singleElementSpan`.

cdumez replied:
Ok.

darinadler commented on Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
```diff
@@ -266,6 +264,14 @@ static void clearNoSenderNotifications(mach_port_t port)
 }
 
+template<typename descriptorType>
+static descriptorType& getDescriptorAndAdvance(std::span<uint8_t>& data)
```
I think we want a different verb other than _get_ here. Maybe _pop_?

darinadler commented on Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
```diff
@@ -424,48 +425,47 @@ static void clearNoSenderNotifications(mach_port_t port)
-        mach_msg_descriptor_t* descriptor = reinterpret_cast<mach_msg_descriptor_t*>(descriptorData);
-        ASSERT(descriptor->type.type == MACH_MSG_PORT_DESCRIPTOR);
-        if (descriptor->type.type != MACH_MSG_PORT_DESCRIPTOR)
+        auto& descriptor = spanReinterpretCast<mach_msg_port_descriptor_t>(remaining.first(sizeof(mach_msg_port_descriptor_t)))[0];
```
This idiom with `first`, `spanReinterpretCast`, and `[0]` needs a wrapper function to make it look cleaner and avoid repeating the type twice. Maybe involving the word _head_ or _front_?

cdumez replied:
Proposing:
```
template<typename T, std::size_t Extent>
const T& reinterpretCastSpanStartTo(std::span<const uint8_t, Extent> span)
{
    return spanReinterpretCast<const T>(span.first(sizeof(T)))[0];
}

template<typename T, std::size_t Extent>
T& reinterpretCastSpanStartTo(std::span<uint8_t, Extent> span)
{
    return spanReinterpretCast<T>(span.first(sizeof(T)))[0];
}
```

darinadler commented on Tools/TestWebKitAPI/Tests/IPC/StreamConnectionEncoderTests.cpp:
```diff
@@ -50,7 +50,7 @@ TEST(StreamConnectionEncoderTest, MinimumMessageSizeCanEncodeExpectedMessageAtAn
     size_t minimumMessageSize = IPC::StreamConnectionEncoder::minimumMessageSize;
     for (size_t i = 0; i < maxAlignment; i += IPC::StreamConnectionEncoder::messageAlignment) {
-        IPC::StreamConnectionEncoder encoder { IPC::MessageName::SetStreamDestinationID, buffer + i,  minimumMessageSize };
+        IPC::StreamConnectionEncoder encoder { IPC::MessageName::SetStreamDestinationID, unsafeMakeSpan(buffer + i,  minimumMessageSize) };
```
An extra space here after the comma (not new).

cdumez committed:
Further reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in Source/WebKit
https://bugs.webkit.org/show_bug.cgi?id=282524

Reviewed by Geoffrey Garen.

* Source/WebKit/GPUProcess/EntryPoint/Cocoa/XPCService/GPUServiceEntryPoint.mm:
(GPU_SERVICE_INITIALIZER):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportReceiveStreamCocoa.mm:
(WebKit::NetworkTransportReceiveStream::receiveLoop):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Platform/IPC/DaemonCoders.cpp:
(WebKit::Daemon::Coder<WebCore::CertificateInfo>::encode):
* Source/WebKit/Platform/IPC/DaemonCoders.h:
* Source/WebKit/Platform/IPC/DaemonDecoder.cpp:
* Source/WebKit/Platform/IPC/Decoder.cpp:
* Source/WebKit/Platform/IPC/Encoder.cpp:
* Source/WebKit/Platform/IPC/Encoder.h:
(IPC::Encoder::encodeObject):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendStream):
(IPC::StreamClientConnection::trySendSyncStream):
(IPC::StreamClientConnection::trySendDestinationIDIfNeeded):
(IPC::StreamClientConnection::sendProcessOutOfStreamMessage):
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.h:
(IPC::StreamConnectionBuffer::header const):
* Source/WebKit/Platform/IPC/StreamConnectionEncoder.h:
* Source/WebKit/Platform/IPC/StreamServerConnection.h:
(IPC::StreamServerConnection::sendSyncReply):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::getDescriptorAndAdvance):
(IPC::Connection::sendOutgoingMessage):
(IPC::createMessageDecoder):
(IPC::Connection::receiveSourceEventHandler):
* Source/WebKit/Platform/IPC/cocoa/DaemonConnectionCocoa.mm:
* Source/WebKit/Platform/IPC/cocoa/MachMessage.h:
(IPC::MachMessage::span):
* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::StreamConnectionEncoder>::Impl::Impl):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionEncoderTests.cpp:
(TestWebKitAPI::TEST(StreamConnectionEncoderTest, MinimumMessageSizeCanEncodeExpectedMessageAtAnyOffset)):

Canonical link: https://commits.webkit.org/286137@main
